### PR TITLE
docs(geo): add geo-js geofences CRUD API docs

### PR DESCRIPTION
**Pull request by TreTuna**
_Description of changes:_
Add docs for Geofences CRUD API operations.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

image

image

image

### CODE REVIEW

**Review by offlineprogrammer — APPROVED** [review]

**Review by thaddmt — APPROVED** [review]

File: src/fragments/lib/geo/js/geofences.mdx [hunk 1]
```diff
@@ -6,9 +6,9 @@
 
 ## Provisioning geofence resources
 
-## Manage Geofences in Your Application
+The first thing needed is to provision a geofence collection resource and configured your app using the instructions in either [Amplify CLI - Geo - Geofence Collection](/cli/geo/geofences) or [Use existing Amazon Location Service resources](/lib/geo/existing-resources) and you have already setup [displaying a map](/lib/geo/maps) in your application.
 
-First, make sure you've provisioned a geofence collection resource and configured your app using the instructions in either [Amplify CLI - Geo - Geofence Collection](/cli/geo/geofences) or [Use existing Amazon Location Service resources](/lib/geo/existing-resources) and you have already setup [displaying a map](/lib/geo/maps) in your application.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:11]: The deleted text looks better and consistent with other pages in Geo.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 2]
```diff
@@ -52,3 +52,328 @@ initializeMap();
 ![A Geofences Management Application](/images/geofence-management-app.png)
 
 ## Geofence API
+
+If you are using a different mapping library or have a need for a more programmatic approach to managing geofences, the `@aws-amplify/geo` package provides methods for managing geofences, but not geofence collections.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:56]:
```suggestion
If you are using a different mapping library or have a need for a programmatic approach to managing geofences, the `@aws-amplify/geo` package provides methods for managing geofences, but not geofence collections.
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 3]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+      [
+        [longitude, latitude],
+        [longitude, latitude],
+        [longitude, latitude],
+        [longitude, latitude]
+      ]
+    ]
+  }
+}
+```
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:110]: I doubt if it's much helpful here.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 4]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+- `options` - optional options object for saving geofences
+  - `collectionName` - the name of the collection to save geofences to.
+    - Defaults to the default collection listed in your `aws-exports.js` file after provisioning a geofence collection resource.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:114]: This should be a continuation of the previous item and not starting a new sublist.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 5]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+```js
+{
+  collectionName: "myCollection";
+}
+```
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:120]: Same, I don't see any value added here.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 6]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+#### Return
+
+The return from `saveGeofences` is a Promise that resolves to an object with both successes and errors for geofences that were successfully created or not.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:124]:
```suggestion
The return from `saveGeofences` is a Promise that resolves to `SaveGeofenceResults` which contains both successes and errors for geofences that were successfully created or failed.
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 7]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+    {
+      error: {
+        code: "string",
+        message: "string",
+      },
+      geofenceId: "string",
+    },
+  ],
+};
+```
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:159]: We should be able to remove this.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 8]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+- `geofenceId` - the geofenceId of the geofence that was saved.
+- `createTime` - the time the geofence was created.
+- `updateTime` - the time the geofence was last updated.
+
+Each error object has the following properties:
+
+- `geofenceId` - the geofenceId of the geofence that failed to be saved.
+- `error` - an error object
+  - `code` - the error code
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:136]: Need to add the potential values for developers to code against. https://docs.aws.amazon.com/location-geofences/latest/APIReference/API_BatchItemError.html ??

File: src/fragments/lib/geo/js/geofences.mdx [hunk 9]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+  });
+} catch (error) {
+  // error handling from logic and validation issues within `createGeofences`
+  throw error;
+}
+
+if (saveGeofenceResults.errors.length > 0) {
+  // error handling that are from the underlying API calls
+  throw new Error();
+}
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:186]: Whether to throw here or not is an application level logic. We should just print out the number of successes and errors to show how to access the response object.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 10]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+        [-123.14695358276366, 49.290090146520434],
+        [-123.1358814239502, 49.294960279811974],
+        [-123.15021514892577, 49.29300108863353],
+        [-123.14909934997559, 49.29132171993048],
+        [-123.14695358276366, 49.290090146520434],
+      ],
+    },
+  });
+} catch (error) {
+  // error handling from logic and validation issues within `createGeofences`
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:179]:
```suggestion
  // errors thrown by input validations of `createGeofence`
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 11]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+if (saveGeofenceResults.errors.length > 0) {
+  // error handling that are from the underlying API calls
+  throw new Error();
+}
+```
+
+### getGeofence
+
+`geoGeofence` is used to get a single geofence object from a collection.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:191]:
```suggestion
`geoGeofence` is used to get a single geofence from a collection.
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 12]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+#### API
+
+```javascript
+Geo.getGeofence(geofenceId, options);
+```
+
+#### Parameters
+
+- `geofenceId` - the `geofenceId` of the geofence to get.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:201]:
```suggestion
- `geofenceId` - the `id` of the geofence to get.
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 13]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+- `options` - optional options object for saving geofences
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:203]:
```suggestion
- `options` - optional options object for getting a geofence
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 14]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+  - `collectionName` - the name of the collection to save geofences to.
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:204]:
```suggestion
  - `collectionName` - the name of the collection to get geofence from.
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 15]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+    - Defaults to the default collection listed in your `aws-exports.js` file after provisioning a geofence collection resource.
+
+```js
+{
+  collectionName: "myCollection";
+}
+```
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:211]:
```suggestion
```

File: src/fragments/lib/geo/js/geofences.mdx [hunk 16]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+
+### saveGeofences
+
+`saveGeofences` is used to save geofences to your collection. It can take a single geofence or an array of geofences.
+
+#### API
+
+```javascript
+Geo.saveGeofences(geofences, options);
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:73]: Show the return type in the API signature.

File: src/fragments/lib/geo/js/geofences.mdx [hunk 17]
```diff
@@ -32,18 +32,18 @@ import maplibregl from "maplibre-gl";
...
+  const map = await createMap({
+    container: "map",
+    center: [-123.1187, 49.2819], // [Longitude, Latitude]
+    zoom: 11,
+  });
 
-    const control = new AmplifyGeofenceControl({})
-    map.addControl(control);
+  const control = new AmplifyGeofenceControl({});
+  map.addControl(control);
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:46]: Are we going to create another PR for adding docs for the geofence control docs?

File: src/fragments/lib/geo/js/geofences.mdx [hunk 18]
```diff
@@ -52,3 +52,328 @@ initializeMap();
...
+```
+
+### getGeofence
+
+`geoGeofence` is used to get a single geofence object from a collection.
+
+#### API
+
+```javascript
+Geo.getGeofence(geofenceId, options);
```
Comment (Amplifiyer) [inline src/fragments/lib/geo/js/geofences.mdx:196]: All of the APIs need to show a return type.

**Review by Amplifiyer — COMMENTED** [review]
Reviewed first couple APIs, the comments for them will be applicable to all APIs here.

File: src/fragments/lib/geo/js/geofences.mdx (hunk 17 above)
Comment (thaddmt) [reply src/fragments/lib/geo/js/geofences.mdx:46]: Created here - https://github.com/aws-amplify/docs/pull/3911

**Review by Amplifiyer — APPROVED** [review]

**Pull request merged**